File: app/src/main/java/com/elementary/tasks/navigation/fragments/MapFragment.java
package com.elementary.tasks.navigation.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.elementary.tasks.R;
import com.elementary.tasks.core.data.models.Place;
import com.elementary.tasks.core.data.models.Reminder;
import com.elementary.tasks.core.fragments.AdvancedMapFragment;
import com.elementary.tasks.core.interfaces.MapCallback;
import com.elementary.tasks.core.utils.MeasureUtils;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.core.view_models.reminders.ActiveGpsRemindersViewModel;
import com.elementary.tasks.databinding.BottomSheetLayoutBinding;
import com.elementary.tasks.databinding.FragmentEventsMapBinding;
import com.elementary.tasks.places.google.LocationPlacesAdapter;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.material.bottomsheet.BottomSheetBehavior;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class MapFragment extends BaseNavigationFragment {

    private FragmentEventsMapBinding binding;
    private ActiveGpsRemindersViewModel viewModel;
    private LocationPlacesAdapter mAdapter = new LocationPlacesAdapter();

    private AdvancedMapFragment mGoogleMap;
    private RecyclerView mEventsList;
    private LinearLayout mEmptyItem;

    private int clickedPosition;
    private int pointer;
    private boolean isDataShowed;

    private MapCallback mReadyCallback = new MapCallback() {
        @Override
        public void onMapReady() {
            mGoogleMap.setSearchEnabled(false);
            showData();
        }
    };
    private BottomSheetBehavior.BottomSheetCallback mSheetCallback = new BottomSheetBehavior.BottomSheetCallback() {
        @Override
        public void onStateChanged(@NonNull View bottomSheet, int newState) {

        }

        @Override
        public void onSlide(@NonNull View bottomSheet, float slideOffset) {

        }
    };
    private GoogleMap.OnMarkerClickListener mOnMarkerClick = new GoogleMap.OnMarkerClickListener() {
        @Override
        public boolean onMarkerClick(Marker marker) {
            mGoogleMap.moveCamera(marker.getPosition(), 0, 0, 0, MeasureUtils.dp2px(getContext(), 192));
            return false;
        }
    };

    private void showClickedPlace(int position, Reminder reminder) {
        int maxPointer = reminder.getPlaces().size() - 1;
        if (position != clickedPosition) {
            pointer = 0;
        } else {
            if (pointer == maxPointer) {
                pointer = 0;
            } else {
                pointer++;
            }
        }
        clickedPosition = position;
        Place place = reminder.getPlaces().get(pointer);
        mGoogleMap.moveCamera(new LatLng(place.getLatitude(), place.getLongitude()), 0, 0, 0, MeasureUtils.dp2px(getContext(), 192));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentEventsMapBinding.inflate(inflater, container, false);
        initMap();
        initViews();
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViewModel();
    }

    private void initViewModel() {
        viewModel = ViewModelProviders.of(this).get(ActiveGpsRemindersViewModel.class);
        viewModel.events.observe(this, reminders -> {
            if (reminders != null && mGoogleMap != null) {
                showData();
            }
        });
    }

    private void initMap() {
        mGoogleMap = AdvancedMapFragment.newInstance(false, false, false, false, false, false,
                ThemeUtil.getInstance(getContext()).isDark());
        mGoogleMap.setCallback(mReadyCallback);
        mGoogleMap.setOnMarkerClick(mOnMarkerClick);
        getFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, mGoogleMap)
                .addToBackStack(null)
                .commit();
    }

    private void initViews() {
        BottomSheetLayoutBinding bottomSheet = binding.bottomSheet;
        mEventsList = bottomSheet.recyclerView;
        mEventsList.setLayoutManager(new LinearLayoutManager(getContext()));

        mAdapter.setActionsListener((view, position, reminder, actions) -> {
            switch (actions) {
                case OPEN:
                    showClickedPlace(position, reminder);
                    break;
            }
        });
        mEventsList.setAdapter(mAdapter);

        mEmptyItem = bottomSheet.emptyItem;
        binding.sheetLayout.setBackgroundColor(ThemeUtil.getInstance(getContext()).getCardStyle());
        BottomSheetBehavior mBottomSheetBehavior = BottomSheetBehavior.from(binding.sheetLayout);
        mBottomSheetBehavior.setBottomSheetCallback(mSheetCallback);
    }

    @Override
    public void onResume() {
        super.onResume();
        if (getCallback() != null) {
            getCallback().onTitleChange(getString(R.string.map));
            getCallback().onFragmentSelect(this);
        }
    }

    private void showData() {
        List<Reminder> data = viewModel.events.getValue();
        if (isDataShowed || data == null) {
            return;
        }
        mAdapter.setData(data);
        boolean mapReady = false;
        for (Reminder reminder : data) {
            for (Place place : reminder.getPlaces()) {
                mapReady = mGoogleMap.addMarker(new LatLng(place.getLatitude(), place.getLongitude()),
                        place.getName(), false, place.getMarker(), false, place.getRadius());
                if (!mapReady) {
                    break;
                }
            }
            if (!mapReady) {
                break;
            }
        }
        isDataShowed = mapReady;
        reloadView();
    }

    private void reloadView() {
        RecyclerView.Adapter adapter = mEventsList.getAdapter();
        int size = 0;
        if (adapter != null) {
            size = mEventsList.getAdapter().getItemCount();
        }
        if (size > 0) {
            mEventsList.setVisibility(View.VISIBLE);
            mEmptyItem.setVisibility(View.GONE);
        } else {
            mEventsList.setVisibility(View.GONE);
            mEmptyItem.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/elementary/tasks/navigation/fragments/MapFragment.kt
package com.elementary.tasks.navigation.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.recyclerview.widget.LinearLayoutManager
import com.elementary.tasks.R
import com.github.naz013.domain.Reminder
import com.elementary.tasks.core.interfaces.ActionsListener
import com.github.naz013.feature.common.android.dp2px
import com.elementary.tasks.core.utils.ListActions
import com.elementary.tasks.databinding.FragmentEventsMapBinding
import com.elementary.tasks.navigation.toolbarfragment.BaseToolbarFragment
import com.elementary.tasks.places.google.LocationPlacesAdapter
import com.elementary.tasks.reminder.lists.active.ActiveGpsRemindersViewModel
import com.elementary.tasks.simplemap.SimpleMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.bottomsheet.BottomSheetBehavior
import org.koin.android.ext.android.get
import org.koin.androidx.viewmodel.ext.android.viewModel

class MapFragment : BaseToolbarFragment<FragmentEventsMapBinding>() {

  private val viewModel by viewModel<ActiveGpsRemindersViewModel>()
  private val mAdapter = LocationPlacesAdapter(get())

  private var simpleMapFragment: SimpleMapFragment? = null
  private var behaviour: BottomSheetBehavior<LinearLayout>? = null

  private var clickedPosition: Int = 0
  private var pointer: Int = 0
  private var isDataShowed: Boolean = false

  override fun inflate(
    inflater: LayoutInflater,
    container: ViewGroup?,
    savedInstanceState: Bundle?
  ) = FragmentEventsMapBinding.inflate(inflater, container, false)

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    behaviour = BottomSheetBehavior.from(binding.placesListCard)
    initMap()
    initViews()
    initViewModel()
  }

  private fun initViewModel() {
    viewModel.events.observe(viewLifecycleOwner) { reminders ->
      if (reminders != null && simpleMapFragment != null) {
        showData(reminders)
      }
    }
  }

  private fun initMap() {
    val map = SimpleMapFragment.newInstance(
      SimpleMapFragment.MapParams(
        isTouch = false,
        isPlaces = false,
        isSearch = false,
        isStyles = false,
        isLayers = true,
        isRadius = false,
        mapStyleParams = SimpleMapFragment.MapStyleParams(
          mapType = prefs.mapType,
          mapStyle = prefs.mapStyle
        )
      )
    )

    map.mapCallback = object : SimpleMapFragment.DefaultMapCallback() {
      override fun onMapReady() {
        super.onMapReady()
        viewModel.events.value?.also { showData(it) }
      }
    }
    map.setOnMarkerClick { marker ->
      simpleMapFragment?.moveCamera(
        pos = marker.position,
        paddingLeft = 0,
        paddingTop = 0,
        paddingRight = 0,
        paddingBottom = dp2px(192)
      )
      false
    }

    childFragmentManager.beginTransaction()
      .replace(R.id.fragment_container, map)
      .addToBackStack(null)
      .commit()

    simpleMapFragment = map
  }

  private fun initViews() {
    binding.recyclerView.layoutManager = LinearLayoutManager(context)
    mAdapter.actionsListener = object : ActionsListener<Reminder> {
      override fun onAction(view: View, position: Int, t: Reminder?, actions: ListActions) {
        when (actions) {
          ListActions.OPEN -> if (t != null) showClickedPlace(position, t)
          else -> {
          }
        }
      }
    }
    binding.recyclerView.adapter = mAdapter
    reloadView()
  }

  private fun showClickedPlace(position: Int, reminder: Reminder) {
    behaviour?.state = BottomSheetBehavior.STATE_COLLAPSED
    val maxPointer = reminder.places.size - 1
    if (position != clickedPosition) {
      pointer = 0
    } else {
      if (pointer == maxPointer) {
        pointer = 0
      } else {
        pointer++
      }
    }
    clickedPosition = position
    val place = reminder.places[pointer]
    simpleMapFragment?.moveCamera(
      LatLng(place.latitude, place.longitude),
      0,
      0,
      0,
      dp2px(192)
    )
  }

  override fun getTitle(): String = getString(R.string.map)

  private fun showData(data: List<Reminder>) {
    val map = simpleMapFragment
    if (isDataShowed || map == null) {
      return
    }
    mAdapter.setData(data)
    if (simpleMapFragment?.isMapReady == true) {
      for (reminder in data) {
        for (place in reminder.places) {
          map.addMarker(
            latLng = LatLng(place.latitude, place.longitude),
            title = place.name,
            markerStyle = place.marker,
            radius = place.radius,
            clear = false,
            animate = false
          )
        }
      }
      isDataShowed = true
    }

    reloadView()
  }

  override fun canGoBack(): Boolean {
    return simpleMapFragment?.onBackPressed() == true
  }

  private fun reloadView() {
    if (mAdapter.itemCount > 0) {
      binding.recyclerView.visibility = View.VISIBLE
      binding.emptyItem.visibility = View.GONE
    } else {
      binding.recyclerView.visibility = View.GONE
      binding.emptyItem.visibility = View.VISIBLE
    }
  }
}
